### PR TITLE
captains games: add 'community picked teams' mode

### PR DESCRIPTION
**Pull request by clifffrey**
In this mode, players of the game itself must select a good player picking order in order to create fair teams.  Every player that wants to play in the game is given the opportunity to build a stack-rank/pick-order of all players that want to play. Once at least 5 players have done so (and at least 30% of all players have done so) it is possible for the referee to start the game.

Note that whenever the list of players changes, it is necessary for everyone to re-confirm their picking order. Due to this, I have removed the "I do not want to play" button when in community-pick mode (i.e. once you volunteer to play, you cannot undo this).  Otherwise it would be too annoying when people toggled the button.

The algorithm for picking is the same 1-2-2-2-... picking order. For each pick, we treat the lists as a ranked-choice vote.  Basically look at everyone's first choice. If one candidate has >50% of the vote, then they are picked. Otherwise, we remove the candidate with the fewest votes from being eligible to be picked this round and start over. Eventually, it is guaranteed that one player will have >50% of the vote. They are then picked for a team, and then we repeat until all players have been picked.

For now, the first-pick to each team becomes "Captain", but this is just for the purposes of dealing with late joiner picking rounds.  Ideally this would change in the UI to not be so loud/clear about who is "captain" (or, we'd make late joiner picking rounds also be "community done", but that would be even more work).

Given that all players are picking at the same time, in parallel, it should be difficult for players to somehow conspire to impact which players will be on their team.

There are some additional (mostly unrelated) changes in here as well.
1) The tournament UI is now always in "collapsed" form. I think that this form is simpler/easier, as it involves fewer nested scrollbars. I also think that it makes it easier for players to figure out how to join, as "escape" won't automatically close the join window.
2) I changed "I want to be a PLAYER" to "I want to play"
3) I removed the old "balanced random teams" game mode, as it was rarely used, and was difficult to actually get balanced teams out of it.
4) change text around the extra-player-info to make it clear that it might be more-widely-visible than just captains seeing it.

cojito (and many others) provided some of the good ideas/encouragement to make this game mode happen.

### Tested Changes:
- [x] I've tested the changes locally or with people.
- [ ] I've not tested the changes.

### CODE REVIEW

File: comfy_panel/special_games/captain.lua [hunk 1]
```diff
@@ -2316,6 +2277,97 @@ function Public.update_captain_player_gui(player, frame)
...
+                        caption = 'Pick Next',
+                        tooltip = 'Add player to the pick order',
+                    })
+                    add_common_cols(tab, player_name)
+                end
+            end
+            flow.captain_community_pick_confirm.enabled = not any_unpicked_players
+            flow.captain_community_pick_confirm.visible = not special.communityPicksConfirmed[player.name]
+        else
+            flow.visible = false
```
Comment (RedRafe) [inline comfy_panel/special_games/captain.lua:2367]: This does not make much sense to recast as false after checking previously if it was true?

File: comfy_panel/special_games/captain.lua [hunk 2]
```diff
@@ -944,33 +948,51 @@ local function start_picking_phase()
...
+        if not picks then
+            force_end_captain_event()
+            return
+        end
+        for i, team in ipairs(picks) do
+            local force_name = i == 1 and 'north' or 'south'
+            game.print(
+                string_format(
+                    'Player %s is assigned captain of team %s - this is just for latejoiner picking purposes. Change captain with /replaceCaptain* commands.',
+                    team[1],
```
Comment (RedRafe) [inline comfy_panel/special_games/captain.lua:980]: Double checking, can you confirm this is `team[1]` always?

File: comfy_panel/special_games/captain.lua [hunk 3]
```diff
@@ -1006,25 +1028,34 @@ local function start_picking_phase()
     Public.update_all_captain_player_guis()
 end
 
-local function check_if_right_number_of_captains(firstRun, referee)
+---@param referee LuaPlayer
+---@return boolean
+local function check_if_right_number_of_captains(referee)
```
Comment (RedRafe) [inline comfy_panel/special_games/captain.lua:1033]: not sure why this method has to take a player as input, would maybe convert it to just type bool function with no inputs and then only print/do stuff based on its result

**Review by RedRafe — COMMENTED** [review]
I like that you changed `string:find` with `starts_with`! Much easier to understand. It's not 100% the scope of this PR, but maybe I would move helper methods to Captain utils or even better, to `utils/table.lua` and `utils/string.lua`. There are also many duplicates with customs table_remove_element and table_contains that can be better unified under utils IMO. Some math.X and table.X not using locale versions but not a biggie again.


Curiosity about the captain_tournament_frame: so other than making the collapsed version as default, the possibility to use the expanded version gets also completely removed and all the code of it deleted as well?

Nice moving `cpt_player_gui_header` to a separated function

**Comment by clifffrey** [comment]
> I like that you changed `string:find` with `starts_with`! Much easier to understand. It's not 100% the scope of this PR, but maybe I would move helper methods to Captain utils or even better, to `utils/table.lua` and `utils/string.lua`. There are also many duplicates with customs table_remove_element and table_contains that can be better unified under utils IMO. Some math.X and table.X not using locale versions but not a biggie again.

I agree that many of these could be cleaned up...

> Curiosity about the captain_tournament_frame: so other than making the collapsed version as default, the possibility to use the expanded version gets also completely removed and all the code of it deleted as well?

Yes, I believe so.  I am excited to have less code to maintain and to avoid needing to test multiple UI configurations.

File: comfy_panel/special_games/captain.lua (hunk 2 above)
Comment (clifffrey) [reply comfy_panel/special_games/captain.lua:980]: I'm not 100% sure what you mean.  `team[1]` will always have the "first pick" for that team, which I think is what we want.

File: comfy_panel/special_games/captain.lua [hunk 4]
```diff
@@ -944,33 +948,51 @@ local function start_picking_phase()
...
+            local force_name = i == 1 and 'north' or 'south'
+            game.print(
+                string_format(
+                    'Player %s is assigned captain of team %s - this is just for latejoiner picking purposes. Change captain with /replaceCaptain* commands.',
+                    team[1],
+                    force_name
+                ),
+                Color.cyan
+            )
+            table.insert(special.captainList, team[1])
```
[the remark this answers is not in the document]
Comment (clifffrey) [reply comfy_panel/special_games/captain.lua]: Done

File: comfy_panel/special_games/captain.lua [hunk 5]
```diff
@@ -1126,25 +1157,32 @@ local function on_gui_switch_state_changed(event)
...
-            insert(special.playerBuckets[1], player)
+    if name == 'captain_community_picking_mode' then
+        special.communityPickingMode = element.switch_state == 'left'
+        special.communityPicksConfirmed = {}
+        special.communityPickOrder = {}
+        special.captainList = {}
+        special.captainGroupAllowed = false
+        if special.test_mode then
+            for _, player in pairs(special.listPlayers) do
+                if math.random() < 0.5 then
```
[the remark this answers is not in the document]
Comment (clifffrey) [reply comfy_panel/special_games/captain.lua]: Done

File: comfy_panel/special_games/captain.lua [hunk 6]
```diff
@@ -2210,19 +2154,33 @@ function Public.update_captain_player_gui(player, frame)
...
-                cpt_volunteers.caption = 'Captain volunteers ('
-                    .. #special.captainList
-                    .. '): '
-                    .. pretty_print_player_list(special.captainList)
                 rem.visible = false
+                if special.communityPickingMode then
+                    confirmed_picks_label.visible = true
+                    local confirmed_pick_players = {}
+                    for player, _ in pairs(special.communityPicksConfirmed) do
+                        table.insert(confirmed_pick_players, player)
```
[the remark this answers is not in the document]
Comment (clifffrey) [reply comfy_panel/special_games/captain.lua]: Done

File: comfy_panel/special_games/captain.lua (hunk 1 above)
Comment (clifffrey) [reply comfy_panel/special_games/captain.lua]: Agreed - Done.

File: comfy_panel/special_games/captain.lua (hunk 3 above)
Comment (clifffrey) [reply comfy_panel/special_games/captain.lua]: Done

File: comfy_panel/special_games/captain.lua [hunk 7]
```diff
@@ -1126,25 +1157,32 @@ local function on_gui_switch_state_changed(event)
...
+        special.captainList = {}
+        special.captainGroupAllowed = false
+        if special.test_mode then
+            for _, player in pairs(special.listPlayers) do
+                if math.random() < 0.5 then
+                    special.communityPicksConfirmed[player] = true
+                    local pick_order = table.deepcopy(special.listPlayers)
+                    -- randomize "pick_order"
+                    for i = #pick_order, 2, -1 do
+                        local j = math.random(i)
```
[the remark this answers is not in the document]
Comment (clifffrey) [reply comfy_panel/special_games/captain.lua]: Done

File: comfy_panel/special_games/captain.lua [hunk 8]
```diff
@@ -1006,25 +1028,34 @@ local function start_picking_phase()
...
     else
         referee.print('Too many captains! Remove some first!', Color.cyan)
+        return false
     end
 end
 
+---@return boolean
+local function have_enough_community_picks()
+    local special = global.special_games_variables.captain_mode
+    local num_confirmed = table.size(special.communityPicksConfirmed)
```
[the remark this answers is not in the document]
Comment (clifffrey) [reply comfy_panel/special_games/captain.lua]: Done

**Review by RedRafe — APPROVED** [review]
Aight, that's a wrap for me. If you feel like tidying up the utils and maybe write a quick comment explanation of the picking algo that would be great, overall this is a thumbs up for me. Curious to try this one out and have it do the job for us. Thanks! :)

**Comment by clifffrey** [comment]
> Aight, that's a wrap for me. If you feel like tidying up the utils and maybe write a quick comment explanation of the picking algo that would be great, overall this is a thumbs up for me. Curious to try this one out and have it do the job for us. Thanks! :)

I added a comment explaining the picking algorithm.  Hopefully 50% of what you wanted is a good start.

**Pull request merged**